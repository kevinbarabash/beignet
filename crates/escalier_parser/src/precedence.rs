use lazy_static::lazy_static;
use std::collections::hash_map::HashMap;

pub type Precedence = u32;

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Associativity {
    Left,
    Right,
    Neither, // indicates a unary operator
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub enum Operator {
    // 18
    // Grouping, - parsed as an atom

    // 17
    MemberAccess,
    OptionalChaining,
    ComputedMemberAccess,
    NewWithArgumentList,
    FunctionCall,
    TemplateLiteral,

    // 16
    // NewWithoutArgList,

    // 15
    // PostfixIncrement,
    // PostfixDecrement,

    // 14
    LogicalNot,
    // BitwiseNot,
    UnaryPlus,
    UnaryMinus,
    // PrefixIncrement,
    // PrefixDecrement,
    Typeof,
    Void,
    Delete,
    Await,
    Throw,

    // 13
    Exponentiation,

    // 12
    Multiplication,
    Division,
    Remainder,

    // 11
    Addition,
    Subtraction,

    // 10
    // BitwiseLeftShift,
    // BitwiseRightShift,
    // BitwiseUnsignedRightShift,

    // 9
    LessThan,
    LessThanOrEqual,
    GreaterThan,
    GreaterThanOrEqual,
    In,
    Instanceof,

    // 8
    Equals,    // always strict
    NotEquals, // always strict
    // StrictEquals,
    // StrictNotEquals,

    // 7
    // BitwiseAnd,

    // 6
    // BitwiseXor,

    // 5
    // BitwiseOr,

    // 4
    LogicalAnd,

    // 3
    LogicalOr,
    NullishCoalescing,

    // 2
    Assignment,
    Conditional,
    // Arrow, - superseded by `fn`
    Yield,
    YieldStar,
    Spread,
    // 1
    // CommaSequence,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct OpInfo {
    prec: Precedence,
    assoc: Associativity,
}

impl OpInfo {
    pub fn new_prefix(prec: Precedence) -> Self {
        Self {
            prec,
            assoc: Associativity::Neither,
        }
    }

    pub fn new_infix(prec: Precedence, assoc: Associativity) -> Self {
        Self { prec, assoc }
    }

    pub fn new_postfix(prec: Precedence) -> Self {
        Self {
            prec,
            assoc: Associativity::Neither,
        }
    }

    pub fn normalized_prec(&self) -> Precedence {
        self.prec * 10
    }

    pub fn infix_postfix_prec(&self) -> Precedence {
        let normalized = self.normalized_prec();
        match self.assoc {
            Associativity::Left => normalized,
            Associativity::Right => normalized - 1,
            Associativity::Neither => normalized + 1,
        }
    }
}

lazy_static! {
    pub static ref PRECEDENCE_TABLE: HashMap<Operator, OpInfo> = {
        let mut table: HashMap<Operator, OpInfo> = HashMap::new();

        table.insert(
            Operator::MemberAccess,
            OpInfo::new_infix(17, Associativity::Left),
        );
        table.insert(
            Operator::OptionalChaining,
            OpInfo::new_infix(17, Associativity::Left),
        );
        table.insert(Operator::ComputedMemberAccess, OpInfo::new_postfix(17));
        table.insert(Operator::NewWithArgumentList, OpInfo::new_postfix(17));
        table.insert(Operator::FunctionCall, OpInfo::new_postfix(17));
        table.insert(Operator::TemplateLiteral, OpInfo::new_postfix(17));

        table.insert(Operator::LogicalNot, OpInfo::new_prefix(14));
        table.insert(Operator::UnaryPlus, OpInfo::new_prefix(14));
        table.insert(Operator::UnaryMinus, OpInfo::new_prefix(14));
        table.insert(Operator::Typeof, OpInfo::new_prefix(14));
        table.insert(Operator::Void, OpInfo::new_prefix(14));
        table.insert(Operator::Delete, OpInfo::new_prefix(14));
        table.insert(Operator::Await, OpInfo::new_prefix(14));
        table.insert(Operator::Throw, OpInfo::new_prefix(14));

        table.insert(
            Operator::Exponentiation,
            OpInfo::new_infix(13, Associativity::Right),
        );

        table.insert(
            Operator::Multiplication,
            OpInfo::new_infix(12, Associativity::Left),
        );
        table.insert(
            Operator::Division,
            OpInfo::new_infix(12, Associativity::Left),
        );
        table.insert(
            Operator::Remainder,
            OpInfo::new_infix(12, Associativity::Left),
        );

        table.insert(
            Operator::Addition,
            OpInfo::new_infix(11, Associativity::Left),
        );
        table.insert(
            Operator::Subtraction,
            OpInfo::new_infix(11, Associativity::Left),
        );

        table.insert(
            Operator::LessThan,
            OpInfo::new_infix(9, Associativity::Left),
        );
        table.insert(
            Operator::LessThanOrEqual,
            OpInfo::new_infix(9, Associativity::Left),
        );
        table.insert(
            Operator::GreaterThan,
            OpInfo::new_infix(9, Associativity::Left),
        );
        table.insert(
            Operator::GreaterThanOrEqual,
            OpInfo::new_infix(9, Associativity::Left),
        );
        table.insert(Operator::In, OpInfo::new_infix(9, Associativity::Left));
        table.insert(
            Operator::Instanceof,
            OpInfo::new_infix(9, Associativity::Left),
        );

        table.insert(Operator::Equals, OpInfo::new_infix(8, Associativity::Left));
        table.insert(
            Operator::NotEquals,
            OpInfo::new_infix(8, Associativity::Left),
        );

        table.insert(
            Operator::LogicalAnd,
            OpInfo::new_infix(4, Associativity::Left),
        );

        table.insert(
            Operator::LogicalOr,
            OpInfo::new_infix(3, Associativity::Left),
        );
        table.insert(
            Operator::NullishCoalescing,
            OpInfo::new_infix(3, Associativity::Left),
        );

        table.insert(
            Operator::Assignment,
            OpInfo::new_infix(2, Associativity::Neither),
        );
        table.insert(
            Operator::Conditional,
            OpInfo::new_infix(2, Associativity::Left),
        );
        table.insert(Operator::Yield, OpInfo::new_infix(2, Associativity::Right));
        table.insert(
            Operator::YieldStar,
            OpInfo::new_infix(2, Associativity::Right),
        );
        table.insert(Operator::Spread, OpInfo::new_infix(2, Associativity::Right));

        table
    };
}
