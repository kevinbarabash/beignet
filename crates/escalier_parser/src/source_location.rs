use std::cmp::{max, min};
use std::ops::Range;

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct Position {
    pub line: usize,
    pub column: usize,
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct SourceLocation {
    pub start: Position,
    pub end: Position,
}

pub type Span = Range<usize>;

pub const DUMMY_SPAN: Span = Span { start: 0, end: 0 };

pub fn merge_spans(left: &Span, right: &Span) -> Span {
    Span {
        start: min(left.start, right.start),
        end: max(left.end, right.end),
    }
}
