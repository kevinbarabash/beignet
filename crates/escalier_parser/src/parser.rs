use std::iter::{Iterator, Peekable};

use crate::source_location::*;
use crate::token::{Token, TokenKind};

pub struct Parser {
    tokens: Peekable<Box<dyn Iterator<Item = Token>>>,
}

impl Parser {
    pub fn new(tokens: Box<dyn Iterator<Item = Token>>) -> Self {
        Self {
            tokens: tokens.peekable(),
        }
    }

    pub fn next(&mut self) -> Token {
        self.tokens.next().unwrap_or(Token {
            kind: TokenKind::Eof,
            loc: SourceLocation {
                start: Position { line: 0, column: 0 },
                end: Position { line: 0, column: 0 },
            },
        })
    }

    pub fn peek(&mut self) -> Token {
        self.tokens.peek().cloned().unwrap_or(Token {
            kind: TokenKind::Eof,
            loc: SourceLocation {
                start: Position { line: 0, column: 0 },
                end: Position { line: 0, column: 0 },
            },
        })
    }
}
