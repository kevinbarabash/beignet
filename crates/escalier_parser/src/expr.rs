use crate::func_param::FuncParam;
use crate::literal::Literal;
use crate::pattern::Pattern;
use crate::source_location::SourceLocation;
use crate::stmt::Stmt;
use crate::type_ann::TypeAnn;

// TODO: track source location
#[derive(Debug, PartialEq, Eq, Clone)]
pub enum ObjectKey {
    Identifier(String), // TODO: use Identifier
    String(String),
    Number(String),
    Computed(Box<Expr>),
}

// TODO: track source location
#[derive(Debug, PartialEq, Eq, Clone)]
pub enum Prop {
    Shorthand { key: String }, // TODO: use Identifier
    Property { key: ObjectKey, value: Expr },
    // TODO:
    // - method
    // - getter
    // - setter
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub enum PropOrSpread {
    Prop(Prop),
    Spread(Expr),
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub enum ExprOrSpread {
    Expr(Expr),
    Spread(Expr),
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub enum ExprKind {
    Identifier(String),
    Literal(Literal),
    TemplateLiteral {
        parts: Vec<Literal>,
        exprs: Vec<Expr>,
    },
    Object {
        properties: Vec<PropOrSpread>,
    },
    Tuple {
        elements: Vec<ExprOrSpread>,
    },
    Binary {
        left: Box<Expr>,
        op: BinaryOp,
        right: Box<Expr>,
    },
    Unary {
        op: UnaryOp,
        right: Box<Expr>,
    },
    Index {
        left: Box<Expr>,
        right: Box<Expr>,
    },
    Function {
        params: Vec<FuncParam>,
        body: BlockOrExpr,
        type_ann: Option<TypeAnn>, // return type
    },
    Call {
        args: Vec<Expr>,
        callee: Box<Expr>,
    },
    Member {
        object: Box<Expr>,
        property: Box<Expr>,
    },
    OptionalChain {
        base: Box<Expr>,
    },
    IfElse {
        cond: Box<Expr>,
        consequent: Block,
        alternate: Option<Block>,
    },
    Match {
        expr: Box<Expr>,
        arms: Vec<MatchArm>,
    },
    Try {
        body: Block,
        // At least `catch` or `finally` must be present
        catch: Option<CatchClause>,
        finally: Option<Block>,
    },
    Do {
        body: Block,
    },
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct CatchClause {
    pub param: Option<Pattern>,
    pub body: Block,
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct MatchArm {
    pub loc: SourceLocation,
    // pub span: Span,
    pub pattern: Pattern,
    pub guard: Option<Box<Expr>>,
    pub body: BlockOrExpr,
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub enum BlockOrExpr {
    Block(Block),
    Expr(Box<Expr>),
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct Block {
    pub loc: SourceLocation,
    pub stmts: Vec<Stmt>,
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub enum BinaryOp {
    Plus,
    Minus,
    Times,
    Divide,
    Modulo,
    Equals,
    NotEquals,
    LessThan,
    LessThanOrEqual,
    GreaterThan,
    GreaterThanOrEqual,
    Or,
    And,
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub enum UnaryOp {
    Plus,
    Minus,
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct Expr {
    pub kind: ExprKind,
    pub loc: SourceLocation,
}
