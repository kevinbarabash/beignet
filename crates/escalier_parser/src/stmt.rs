use escalier_ast::values::*;

use crate::expr::parse_expression;
use crate::parse_error::ParseError;
use crate::pattern::*;
use crate::type_ann::parse_type_ann;
use crate::util::*;

pub fn parse_statement(
    node: &tree_sitter::Node,
    src: &str,
) -> Result<Option<Statement>, ParseError> {
    let kind = match node.kind() {
        "lexical_declaration" => {
            return parse_declaration(node, false, src);
        }
        "class_declaration" => {
            return parse_class_decl(node, src);
        }
        "expression_statement" => {
            let expr = node.named_child(0).unwrap();
            let expr = parse_expression(&expr, src)?;
            StmtKind::ExprStmt(expr)
        }
        "ambient_declaration" => {
            let decl = node.named_child(0).unwrap();
            return parse_declaration(&decl, true, src);
        }
        "type_alias_declaration" => {
            let name = node.child_by_field_name("name").unwrap();
            let id = Ident {
                loc: SourceLocation::from(&name),
                span: name.byte_range(),
                name: text_for_node(&name, src)?,
            };
            let type_ann = node.child_by_field_name("value").unwrap();
            let type_ann = parse_type_ann(&type_ann, src)?;

            let type_params = parse_type_params_for_node(node, src)?;

            StmtKind::TypeDecl(TypeDecl {
                declare: false,
                id,
                type_ann,
                type_params,
            })
        }
        "for_in_statement" => {
            let left_node = node.child_by_field_name("left").unwrap();
            let right_node = node.child_by_field_name("right").unwrap();
            let body_node = node.child_by_field_name("body").unwrap();

            StmtKind::ForStmt(ForStmt {
                pattern: Box::from(parse_pattern(&left_node, src)?),
                expr: Box::from(parse_expression(&right_node, src)?),
                body: parse_block_statement(&body_node, src)?,
            })
        }
        "return_statement" => {
            let arg = match node.named_child(0) {
                Some(arg) => Some(Box::from(parse_expression(&arg, src)?)),
                None => None,
            };

            StmtKind::ReturnStmt(ReturnStmt { arg })
        }
        "comment" => {
            return Ok(None); // ignore comments
        }
        "ERROR" => {
            return Err(ParseError::from(format!(
                "failed to parse: '{}'",
                text_for_node(node, src)?
            )))
        }
        _ => return Err(ParseError::from(format!("unhandled: {:#?}", node))),
    };

    Ok(Some(Statement {
        loc: SourceLocation::from(node),
        span: node.byte_range(),
        kind,
    }))

    // $.export_statement,
    // $.import_statement,
    // $.debugger_statement,
    // $.expression_statement,
    // $.declaration,
    // $.statement_block,

    // $.if_statement,
    // $.switch_statement,
    // $.for_statement,
    // $.for_in_statement,
    // $.while_statement,
    // $.do_statement,
    // $.try_statement,
    // $.with_statement,

    // $.break_statement,
    // $.continue_statement,
    // $.return_statement,
    // $.throw_statement,
    // $.empty_statement,
    // $.labeled_statement
}

fn parse_declaration(
    node: &tree_sitter::Node,
    declare: bool,
    src: &str,
) -> Result<Option<Statement>, ParseError> {
    if node.has_error() {
        // TODO: get actual error node so that we can report where the error is
        return Err(ParseError::from("Error parsing declaration"));
    }

    let decl = node.child_by_field_name("decl").unwrap();
    let rec = node.child_by_field_name("rec").is_some();

    let name = decl.child_by_field_name("name").unwrap();
    let pattern = parse_pattern(&name, src)?;

    let type_ann = if let Some(type_ann) = decl.child_by_field_name("type") {
        Some(parse_type_ann(&type_ann, src)?)
    } else {
        None
    };

    let kind = if rec {
        // `let fib = fix((fib) => (n) => ...)`
        // TODO: Fix always wraps a lambda

        let value = decl.child_by_field_name("value").unwrap();

        let lambda_expr = Expr {
            loc: SourceLocation::from(&decl),
            span: decl.byte_range(),
            kind: ExprKind::Lambda(Lambda {
                params: vec![EFnParam {
                    pat: pattern.clone(),
                    type_ann: type_ann.clone(),
                    optional: false,
                }],
                body: BlockOrExpr::Expr(Box::from(parse_expression(&value, src)?)),
                is_async: false,
                return_type: None,
                type_params: None, // TODO: support type params on VarDecls
            }),
            inferred_type: None,
        };

        let fix = Expr {
            loc: SourceLocation::from(&decl),
            span: decl.byte_range(),
            kind: ExprKind::Fix(Fix {
                expr: Box::from(lambda_expr),
            }),
            inferred_type: None,
        };

        StmtKind::VarDecl(VarDecl {
            pattern,
            type_ann,
            init: Some(Box::from(fix)),
            declare,
        })
    } else {
        let init = if let Some(init) = decl.child_by_field_name("value") {
            Some(Box::from(parse_expression(&init, src)?))
        } else {
            None
        };

        StmtKind::VarDecl(VarDecl {
            pattern,
            type_ann,
            init,
            declare,
        })
    };

    Ok(Some(Statement {
        loc: SourceLocation::from(node),
        span: node.byte_range(),
        kind,
    }))
}

fn parse_class_decl(node: &tree_sitter::Node, src: &str) -> Result<Option<Statement>, ParseError> {
    if node.has_error() {
        // TODO: get actual error node so that we can report where the error is
        return Err(ParseError::from("Error parsing declaration"));
    }

    let name_node = node.child_by_field_name("name").unwrap();
    let body_node = node.child_by_field_name("body").unwrap();

    let mut class_members: Vec<ClassMember> = vec![];
    let mut cursor = body_node.walk();
    for child in body_node.named_children(&mut cursor) {
        let kind = child.kind();

        eprintln!("child.kind() = {kind}");

        match kind {
            "method_definition" => {
                let name_node = child.child_by_field_name("name").unwrap();
                let key = Ident {
                    loc: SourceLocation::from(&name_node),
                    span: name_node.byte_range(),
                    name: text_for_node(&name_node, src)?,
                };

                let params = parse_formal_parameters(
                    &child.child_by_field_name("parameters").unwrap(),
                    src,
                )?;
                let body_node = child.child_by_field_name("body").unwrap();
                let body = parse_block_statement(&body_node, src)?;
                let return_type = match child.child_by_field_name("return_type") {
                    Some(type_ann) => Some(parse_type_ann(&type_ann, src)?),
                    None => None,
                };

                let kind = match child.child_by_field_name("kind") {
                    Some(kind) => match text_for_node(&kind, src)?.as_str() {
                        "get" => MethodKind::Getter,
                        "set" => MethodKind::Setter,
                        "*" => todo!(),
                        _ => panic!("Invalid method kind"), // TODO: report an error
                    },
                    None => MethodKind::Method,
                };

                // TODO: add support to the AST for these
                let is_static = child.child_by_field_name("static").is_some();
                let is_mutating = child.child_by_field_name("mut").is_some();

                let type_params = parse_type_params_for_node(&child, src)?;

                if key.name.as_str() == "constructor" {
                    class_members.push(ClassMember::Constructor(Constructor { params, body }));
                } else {
                    class_members.push(ClassMember::Method(ClassMethod {
                        key,
                        kind,
                        lambda: Lambda {
                            params,
                            body: BlockOrExpr::Block(body),
                            is_async: child.child_by_field_name("async").is_some(),
                            return_type,
                            type_params,
                        },
                        is_static,
                        is_mutating,
                    }))
                }
            }
            "public_field_definition" => {
                let name_node = child.child_by_field_name("name").unwrap();
                let key = Ident {
                    loc: SourceLocation::from(&name_node),
                    span: name_node.byte_range(),
                    name: text_for_node(&name_node, src)?,
                };

                let value = match child.child_by_field_name("value") {
                    Some(value) => Some(Box::from(parse_expression(&value, src)?)),
                    None => None,
                };
                let type_ann = match child.child_by_field_name("type") {
                    Some(type_ann) => Some(Box::from(parse_type_ann(&type_ann, src)?)),
                    None => None,
                };

                class_members.push(ClassMember::Prop(ClassProp {
                    key,
                    value,
                    type_ann,
                    is_static: child.child_by_field_name("static").is_some(),
                    is_optional: child.child_by_field_name("optional").is_some(),
                    is_mutable: child.child_by_field_name("mut").is_some(),
                }))
            }
            "private_field_definition" => todo!(),
            _ => todo!(),
        }
    }

    let stmt = Statement {
        loc: SourceLocation::from(node),
        span: node.byte_range(),
        kind: StmtKind::ClassDecl(ClassDecl {
            ident: Ident {
                loc: SourceLocation::from(&name_node),
                span: name_node.byte_range(),
                name: text_for_node(&name_node, src)?,
            },
            class: Box::from(Class {
                ident: Ident {
                    loc: SourceLocation::from(&name_node),
                    span: name_node.byte_range(),
                    name: text_for_node(&name_node, src)?,
                },
                body: class_members,
                type_params: parse_type_params_for_node(node, src)?, // TODO
            }),
        }),
    };

    Ok(Some(stmt))
}

pub fn parse_block_statement(node: &tree_sitter::Node, src: &str) -> Result<Block, ParseError> {
    assert_eq!(node.kind(), "statement_block");

    let mut cursor = node.walk();

    let child_count = node.named_child_count();

    let mut stmts: Vec<Statement> = vec![];

    for (i, child) in node.named_children(&mut cursor).into_iter().enumerate() {
        let is_last = i == child_count - 1;
        if is_last {
            // This is the only place where a named `expression` node
            // should exist.  Everywhere else its contents should be
            // inline.
            if child.kind() == "expression" {
                let expr = child.named_child(0).unwrap();
                let expr = parse_expression(&expr, src)?;
                stmts.push(Statement {
                    loc: SourceLocation::from(&child),
                    span: child.byte_range(),
                    kind: StmtKind::ExprStmt(expr),
                })
            } else if let Some(result) = parse_statement(&child, src)? {
                stmts.push(result);
            }
        } else if let Some(result) = parse_statement(&child, src)? {
            stmts.push(result);
        }
    }

    Ok(Block {
        span: node.byte_range(),
        stmts,
    })
}
