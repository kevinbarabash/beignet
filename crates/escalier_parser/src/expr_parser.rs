use crate::expr::{BinaryOp, Expr, ExprKind, UnaryOp};
use crate::parser::Parser;
use crate::precedence::{Associativity, Operator, PRECEDENCE_TABLE};
use crate::source_location::*;
use crate::stmt::Stmt;
use crate::stmt_parser::parse_stmt;
use crate::token::{Token, TokenKind};

fn get_prefix_precedence(op: &Token) -> Option<(u8, Associativity)> {
    match &op.kind {
        TokenKind::Plus => PRECEDENCE_TABLE.get(&Operator::UnaryPlus).cloned(),
        TokenKind::Minus => PRECEDENCE_TABLE.get(&Operator::UnaryMinus).cloned(),
        _ => None,
    }
}

fn get_infix_precedence(op: &Token) -> Option<(u8, Associativity)> {
    match &op.kind {
        // multiplicative
        TokenKind::Times => PRECEDENCE_TABLE.get(&Operator::Multiplication).cloned(),
        TokenKind::Divide => PRECEDENCE_TABLE.get(&Operator::Division).cloned(),
        // TODO: modulo

        // additive
        TokenKind::Plus => PRECEDENCE_TABLE.get(&Operator::Addition).cloned(),
        TokenKind::Minus => PRECEDENCE_TABLE.get(&Operator::Subtraction).cloned(),

        // equality
        TokenKind::Equals => PRECEDENCE_TABLE.get(&Operator::Equals).cloned(),
        TokenKind::LessThan => PRECEDENCE_TABLE.get(&Operator::LessThan).cloned(),
        TokenKind::LessThanOrEqual => PRECEDENCE_TABLE.get(&Operator::LessThanOrEqual).cloned(),
        TokenKind::GreaterThan => PRECEDENCE_TABLE.get(&Operator::GreaterThan).cloned(),
        TokenKind::GreaterThanOrEqual => {
            PRECEDENCE_TABLE.get(&Operator::GreaterThanOrEqual).cloned()
        }

        // logic
        TokenKind::And => PRECEDENCE_TABLE.get(&Operator::LogicalAnd).cloned(),
        TokenKind::Or => PRECEDENCE_TABLE.get(&Operator::LogicalOr).cloned(),

        // assignment
        TokenKind::Assign => PRECEDENCE_TABLE.get(&Operator::Assignment).cloned(),
        _ => None,
    }
}

fn get_postfix_precedence(op: &Token) -> Option<(u8, Associativity)> {
    match &op.kind {
        TokenKind::LeftBracket => PRECEDENCE_TABLE
            .get(&Operator::ComputedMemberAccess)
            .cloned(),
        TokenKind::LeftParen => PRECEDENCE_TABLE.get(&Operator::FunctionCall).cloned(),
        // TODO: handle optional chaining
        TokenKind::Dot => PRECEDENCE_TABLE.get(&Operator::MemberAccess).cloned(),
        _ => None,
    }
}

fn parse_params(parser: &mut Parser) -> Vec<String> {
    let mut params = Vec::new();
    while parser.peek().kind != TokenKind::RightParen {
        let param = parser.next();
        if let TokenKind::Identifier(name) = param.kind {
            params.push(name.to_owned());
        } else {
            panic!("Expected identifier, got {:?}", param);
        }

        match parser.peek().kind {
            TokenKind::RightParen => break,
            TokenKind::Comma => {
                parser.next();
            }
            _ => panic!("Expected comma or right paren, got {:?}", parser.peek()),
        }
    }
    params
}

fn parse_block(parser: &mut Parser) -> Vec<Stmt> {
    let mut stmts = Vec::new();
    while parser.peek().kind != TokenKind::RightBrace {
        stmts.push(parse_stmt(parser));
    }
    stmts
}

fn parse_expr_with_precedence(parser: &mut Parser, precedence: u8) -> Expr {
    let next = parser.next();

    let mut lhs = match &next.kind {
        TokenKind::Number(n) => Expr {
            kind: ExprKind::Number(n.to_owned()),
            loc: next.loc.clone(),
        },
        TokenKind::Identifier(id) => Expr {
            kind: ExprKind::Identifier(id.to_owned()),
            loc: next.loc.clone(),
        },
        TokenKind::LeftParen => {
            let lhs = parse_expr_with_precedence(parser, 0);
            assert_eq!(parser.next().kind, TokenKind::RightParen);
            lhs
        }
        TokenKind::Fn => {
            assert_eq!(parser.next().kind, TokenKind::LeftParen);
            let params = parse_params(parser);
            assert_eq!(parser.next().kind, TokenKind::RightParen);
            assert_eq!(parser.next().kind, TokenKind::LeftBrace);
            let body = parse_block(parser);
            let close_brace = parser.next();
            assert_eq!(close_brace.kind, TokenKind::RightBrace);

            let loc = merge_locations(&next.loc, &close_brace.loc);
            Expr {
                kind: ExprKind::Function { params, body },
                loc,
            }
        }
        t => match get_prefix_precedence(&next) {
            Some(precendence) => {
                let op = match t {
                    TokenKind::Plus => UnaryOp::Plus,
                    TokenKind::Minus => UnaryOp::Minus,
                    _ => panic!("unexpected token: {:?}", t),
                };

                let rhs = parse_expr_with_precedence(parser, precendence.0);
                let loc = merge_locations(&next.loc, &rhs.loc);
                Expr {
                    kind: ExprKind::Unary {
                        op,
                        right: Box::new(rhs),
                    },
                    loc,
                }
            }
            None => panic!("unexpected token: {:?}", t),
        },
    };

    loop {
        let next = parser.peek();
        if let TokenKind::Eof = next.kind {
            break;
        }

        if let Some(next_precedence) = get_postfix_precedence(&next) {
            if precedence >= next_precedence.0 {
                break;
            }

            let precedence = if next_precedence.1 == Associativity::Left {
                next_precedence.0
            } else {
                next_precedence.0 - 1
            };

            parser.next();

            lhs = match &next.kind {
                TokenKind::LeftBracket => {
                    let rhs = parse_expr_with_precedence(parser, 0);
                    let loc = merge_locations(&lhs.loc, &rhs.loc);
                    assert_eq!(parser.next().kind, TokenKind::RightBracket);
                    Expr {
                        kind: ExprKind::Index {
                            left: Box::new(lhs),
                            right: Box::new(rhs),
                        },
                        loc,
                    }
                }
                TokenKind::LeftParen => {
                    let mut args = Vec::new();
                    while parser.peek().kind != TokenKind::RightParen {
                        args.push(parse_expr_with_precedence(parser, 0));

                        match parser.peek().kind {
                            TokenKind::RightParen => break,
                            TokenKind::Comma => {
                                parser.next();
                            }
                            _ => panic!("Expected comma or right paren, got {:?}", parser.peek()),
                        }
                    }
                    let loc = merge_locations(&lhs.loc, &parser.peek().loc);
                    assert_eq!(parser.next().kind, TokenKind::RightParen);
                    Expr {
                        kind: ExprKind::Call {
                            callee: Box::new(lhs),
                            args,
                        },
                        loc,
                    }
                }
                TokenKind::Dot => {
                    let rhs = parse_expr_with_precedence(parser, precedence);
                    let loc = merge_locations(&lhs.loc, &rhs.loc);
                    Expr {
                        kind: ExprKind::Member {
                            object: Box::new(lhs),
                            property: Box::new(rhs),
                        },
                        loc,
                    }
                }
                _ => panic!("unexpected token: {:?}", next),
            };

            continue;
        }

        if let Some(next_precedence) = get_infix_precedence(&next) {
            if precedence >= next_precedence.0 {
                break;
            }

            parser.next();

            let op: BinaryOp = match &next.kind {
                TokenKind::Plus => BinaryOp::Plus,
                TokenKind::Minus => BinaryOp::Minus,
                TokenKind::Times => BinaryOp::Times,
                TokenKind::Divide => BinaryOp::Divide,
                TokenKind::Equals => BinaryOp::Equals,
                TokenKind::NotEquals => BinaryOp::NotEquals,
                TokenKind::LessThan => BinaryOp::LessThan,
                TokenKind::LessThanOrEqual => BinaryOp::LessThanOrEqual,
                TokenKind::GreaterThan => BinaryOp::GreaterThan,
                TokenKind::GreaterThanOrEqual => BinaryOp::GreaterThanOrEqual,
                TokenKind::And => BinaryOp::And,
                TokenKind::Or => BinaryOp::Or,
                _ => panic!("unexpected token: {:?}", next),
            };

            let precedence = if next_precedence.1 == Associativity::Left {
                next_precedence.0
            } else {
                next_precedence.0 - 1
            };

            let rhs = parse_expr_with_precedence(parser, precedence);
            let loc = merge_locations(&lhs.loc, &rhs.loc);

            lhs = Expr {
                kind: ExprKind::Binary {
                    op,
                    left: Box::new(lhs),
                    right: Box::new(rhs),
                },
                loc,
            };

            continue;
        }

        break;
    }

    lhs
}

pub fn parse_expr(parser: &mut Parser) -> Expr {
    parse_expr_with_precedence(parser, 0)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::lexer::Lexer;

    pub fn parse(input: &str) -> Expr {
        let mut lexer = Lexer::new(input);
        let tokens = lexer.lex();
        let mut parser = Parser::new(tokens);
        parse_expr(&mut parser)
    }

    #[test]
    fn parse_simple_addition() {
        insta::assert_debug_snapshot!(parse("1 + 2 + 3"));
    }

    #[test]
    fn parse_addition_and_subtraction() {
        insta::assert_debug_snapshot!(parse("1 + 2 - 3 + 4"));
    }

    #[test]
    fn parse_additive_and_multiplicative() {
        insta::assert_debug_snapshot!(parse("1 * 2 + 3"));
    }

    #[test]
    fn parse_parens() {
        insta::assert_debug_snapshot!(parse("5 * (x + 1)"));
    }

    #[test]
    fn parse_comparisons_and_logic() {
        insta::assert_debug_snapshot!(parse("a > b && c >= d || e < f && g <= h"));
    }

    #[test]
    fn parse_unary_operators() {
        insta::assert_debug_snapshot!(parse("--a - +b"));
    }

    #[test]
    fn parse_indexing() {
        insta::assert_debug_snapshot!(parse("a[1][c]"));
    }

    #[test]
    fn parse_function() {
        let src = r#"fn () { let x = 5; let y = 10; return x + y; }"#;
        insta::assert_debug_snapshot!(parse(src));
    }

    #[test]
    fn parse_function_with_params() {
        let src = r#"fn (x, y) { return x + y; }"#;
        insta::assert_debug_snapshot!(parse(src));
    }

    #[test]
    #[should_panic]
    fn parse_function_expected_comma_or_left_brace() {
        let src = r#"fn (x, y { return x + y; }"#;
        parse(src);
    }

    #[test]
    #[should_panic]
    fn parse_function_expected_identifier() {
        let src = r#"fn (, y) { return x + y; }"#;
        parse(src);
    }

    #[test]
    fn parse_function_call() {
        let src = r#"add(5 * i, 10 * j)"#;
        insta::assert_debug_snapshot!(parse(src));
    }

    #[test]
    fn parse_call_expr() {
        let src = r#"foo[bar](5, 10)"#;
        insta::assert_debug_snapshot!(parse(src));
    }

    #[test]
    fn parse_member_access() {
        insta::assert_debug_snapshot!(parse("a.b.c"));
        insta::assert_debug_snapshot!(parse("a.b+c.d"));
    }
}
