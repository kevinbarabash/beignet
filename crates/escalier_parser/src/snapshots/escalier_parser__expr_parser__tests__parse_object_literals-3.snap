---
source: crates/escalier_parser/src/expr_parser.rs
expression: "parse(\"{ a: 1, b: 2 }\")"
---
Expr {
    kind: Object {
        properties: [
            Prop(
                Property {
                    key: Identifier(
                        "a",
                    ),
                    value: Expr {
                        kind: Literal(
                            Number(
                                "1",
                            ),
                        ),
                        span: 5..6,
                    },
                },
            ),
            Prop(
                Property {
                    key: Identifier(
                        "b",
                    ),
                    value: Expr {
                        kind: Literal(
                            Number(
                                "2",
                            ),
                        ),
                        span: 11..12,
                    },
                },
            ),
        ],
    },
    span: 0..14,
}
